File: solutions/go/minesweeper/minesweeper.go
package minesweeper

const mine = '*'

// Annotate returns an annotated board
func Annotate(board []string) []string {
	// check for no rows or columns
	rows := len(board)
	if rows == 0 {
		return board
	}
	cols := len(board[0])
	if cols == 0 {
		return board
	}

	// function to count neighboring mines for a given row
	checkRow := func(row int, col int) (count int) {
		if row < 0 || row >= rows {
			return
		}
		count += boolToInt(board[row][col] == mine)
		count += boolToInt(col-1 >= 0 && board[row][col-1] == mine)
		count += boolToInt(col+1 < cols && board[row][col+1] == mine)
		return
	}

	result := board
	// Traverse the board to find empty space and count mines around it
	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if board[r][c] == mine {
				continue // skip if cell is a mine
			}

			mines := checkRow(r, c)
			mines += checkRow(r-1, c)
			mines += checkRow(r+1, c)

			// replace rune in this row's string with the number of mines
			if mines > 0 {
				result[r] = result[r][:c] + string(rune(int('0')+mines)) + result[r][c+1:]
			}
		}
	}
	return result
}

func boolToInt(b bool) int {
	if b {
		return 1
	}
	return 0
}

File: solutions/julia/minesweeper/minesweeper.jl
function annotate(input)
    # check for no rows or columns
    if length(input) == 0 || input == [""]
        return input
    end

    # create matrix of minefield
    minefield = permutedims(reduce(hcat, collect.(input)))
    rows = size(minefield)[1]
    cols = size(minefield)[2]
    cells = CartesianIndices(minefield)

    neighbors = [(-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1)]

    # Traverse minefield to find mine and count mines around it
    for cell in cells
        # skip if cell is not a mine
        if minefield[cell] != '*'
            continue
        end

        for offest in neighbors
            neighbor = cell + CartesianIndex(offest)

            # skip if neighbor is not in bounds or is a mine
            if !checkbounds(Bool, minefield, neighbor) || minefield[neighbor] == '*'
                continue
            end

            # otherwise add 1 to neighbor value
            if minefield[neighbor] == ' '
                minefield[neighbor] = 48 # char = '0'
            end
            minefield[neighbor] += 1
        end
    end
    reduce(*, reshape(minefield, (rows, cols)), dims=2)[:]
end
